client/sync-state-rpc/src/lib.rs

```diff
@@ -40,62 +40,87 @@ enum Error<Block: BlockT> {
+// not going to mess with it. This here is a super-hack though and if we do need it, it has to be moved somewhere where
+// it makes sense.
+/// Wraps an owned `String` and implements the `Error` trait for it.
+#[derive(Debug)]
+struct StringError {
+    error: String
+}
+
+impl std::error::Error for StringError {
+    fn description(&self) -> &str {
+        &self.error
+    }
 }
 
-/// The handler for sync state RPC calls.
-pub struct SyncStateRpcHandler<TBl: BlockT, TCl> {
+impl std::fmt::Display for StringError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Error: {}", self.error)
+    }
+}
+
+/// An api for sync state RPC calls.
+pub struct SyncStateRpc<Block: BlockT, Client> {
 	chain_spec: Box<dyn sc_chain_spec::ChainSpec>,
-	client: Arc<TCl>,
-	shared_authority_set: SharedAuthoritySet<TBl>,
-	shared_epoch_changes: SharedEpochChanges<TBl>,
+	client: Arc<Client>,
+	shared_authority_set: SharedAuthoritySet<Block>,
+	shared_epoch_changes: SharedEpochChanges<Block>,
 	deny_unsafe: sc_rpc_api::DenyUnsafe,
 }
 
-impl<TBl, TCl> SyncStateRpcHandler<TBl, TCl>
-	where
-		TBl: BlockT,
-		TCl: HeaderBackend<TBl> + sc_client_api::AuxStore + 'static,
+impl<Block, Client> SyncStateRpc<Block, Client>
+where
+	Block: BlockT,
+	Client: HeaderBackend<Block> + sc_client_api::AuxStore + 'static,
 {
-	/// Create a new handler.
+	/// Create a new sync state RPC helper.
 	pub fn new(
 		chain_spec: Box<dyn sc_chain_spec::ChainSpec>,
-		client: Arc<TCl>,
-		shared_authority_set: SharedAuthoritySet<TBl>,
-		shared_epoch_changes: SharedEpochChanges<TBl>,
+		client: Arc<Client>,
+		shared_authority_set: SharedAuthoritySet<Block>,
+		shared_epoch_changes: SharedEpochChanges<Block>,
 		deny_unsafe: sc_rpc_api::DenyUnsafe,
 	) -> Self {
 		Self {
 			chain_spec, client, shared_authority_set, shared_epoch_changes, deny_unsafe,
 		}
 	}
 
-	fn build_sync_state(&self) -> Result<sc_chain_spec::LightSyncState<TBl>, Error<TBl>> {
+	/// Convert this [`SyncStateRpc`] to a RPC module.
+	pub fn into_rpc_module(self) -> Result<RpcModule<Self>, JsonRpseeError> {
+		let mut module = RpcModule::new(self);
+
+		// Returns the json-serialized chainspec running the node, with a sync state.
+		module.register_method::<jsonrpsee::types::JsonValue, _>("sync_state_genSyncSpec", |params, sync_state| {
+			sync_state.deny_unsafe.check_if_safe()?;
+
+			let raw = params.one()?;
+			let mut chain_spec = sync_state.chain_spec.cloned_box();
+			let current_sync_state = sync_state.build_sync_state()?;
+			chain_spec.set_light_sync_state(current_sync_state.to_serializable());
+			let string = chain_spec.as_json(raw).map_err(|error| {
+				let string_err = StringError { error };
+				jsonrpsee::types::error::CallError::Failed(Box::new(string_err))
+			})?;
+
+			serde_json::from_str(&string).map_err(|json_err| jsonrpsee::types::error::CallError::Failed(Box::new(json_err)))
```

Comment:
you don't have to convert it to a jsonvalue, String should be sufficient the rest is taking care of by jsonrpsee